mycpp/gc_dict.h

```diff
@@ -307,49 +283,92 @@ void Dict<K, V>::clear() {
   len_ = 0;
 }
 
-// Returns the position in the array.  Used by dict_contains(), index(),
-// get(), and set().
-//
-// For now this does a linear search.
 // TODO:
-// - hash functions, and linear probing.
-// - resizing based on load factor
-//   - which requires rehashing (re-insert all items)
 // - Special case to intern Str* when it's hashed?  How?
 //   - Should we have wrappers like:
 //   - V GetAndIntern<V>(D, &string_key)
 //   - SetAndIntern<V>(D, &string_key, value)
 //   This will enable duplicate copies of the string to be garbage collected
 template <typename K, typename V>
-int Dict<K, V>::position_of_key(K key) {
-  for (int i = 0; i < capacity_; ++i) {
-    int special = entry_->items_[i];  // NOT an index now
-    if (special == kDeletedEntry) {
-      continue;  // keep searching
+int Dict<K, V>::hash_and_probe(K key) const {
+  if (capacity_ == 0) {
+    return kNotFound;
+  }
+
+  // Hash the key onto a slot in the index. If the first slot is occupied, probe
+  // until an empty one is found.
+  unsigned h = hash_key(key);
+  int init_bucket = h % index_len_;
+
+  for (int i = 0; i < index_len_; ++i) {
+    // Start at init_bucket and wrap araound
+    int slot = (i + init_bucket) % index_len_;
+
+    int kv_index = index_->items_[slot];
+    DCHECK(kv_index < len_);
+    // Optimistically this is the common case once the table has been populated.
+    if (kv_index >= 0) {
+      unsigned h2 = hash_key(keys_->items_[kv_index]);
+      if (h == h2 && keys_equal(keys_->items_[kv_index], key)) {
+        return slot;
+      }
     }
-    if (special == kEmptyEntry) {
-      return -1;  // not found
+
+    if (kv_index == kEmptyEntry) {
+      // If there isn't room in the entry arrays, tell the caller to resize.
+      return len_ < capacity_ ? slot : kNotFound;
     }
+
+    // Tombstone or collided keys unequal. Keep scanning.
+    DCHECK(kv_index >= 0 || kv_index == kDeletedEntry);
+  }
+
+  return kNotFound;
+}
+
+template <typename K, typename V>
+int Dict<K, V>::find_kv_index(K key) const {
+  if (index_ != nullptr) {
+    // Common case.
+    int pos = hash_and_probe(key);
+    if (pos == kNotFound || index_->items_[pos] < 0) {
+      return kNotFound;
+    }
+    return index_->items_[pos];
+  }
+
+  // GlobalDict. Just scan.
+  for (int i = 0; i < len_; ++i) {
     if (keys_equal(keys_->items_[i], key)) {
       return i;
     }
   }
-  return -1;  // table is completely full?  Does this happen?
+
+  // Not found.
+  return kNotFound;
 }
 
 template <typename K, typename V>
 void Dict<K, V>::set(K key, V val) {
-  int pos = position_of_key(key);
-  if (pos == -1) {  // new pair
-    reserve(len_ + 1);
+  DCHECK(obj_header().heap_tag != HeapTag::Global);
+  int pos = hash_and_probe(key);
+  if (pos == kNotFound) {
+    // No room. Resize and see if we can find a slot.
+    reserve(RoundCapacity(capacity_ * 2));
```

Comment:
Ah yeah. I was just using it here to handle the `capacity_ == 0`  case from [this thread](https://github.com/oilshell/oil/pull/1729#discussion_r1333843296). I'm fine dropping it though

Reply:
Oh yeah the intention behind reserve() is that you call it with the exact number of things you need, and then it rounds up if necessary, based on an internal policy.

So we used to call it with reserve(n+1) when there was no key.  But it is valid for it to use  2*n behind the scenes -- that's hidden from the caller.

So 90% of the time, it should be a no-op, because it's already big enough

I think vector::reserve() works the same way


I think that logic is sound because it puts all the policy in one place.

The "attempts" logic was weird because reserve() is supposed to GUARANTEE you have enough space for that number of items.

It may be worth adding some comments/tests on that  ... I think we  are going to end up tuning how it works

Reply:
yeah and probably calling the param `num_desired` is better than `n`, to make that clear

It ensures you have enough space for that number of items, but it can also allocate more, in order to save resizings

It is supposed to completely encapsulate the growth policy, like std::vector::reserve()

Reply:
Ah missed these! I'll replace the thing i pushed with `capacity_ + 1`

Reply:
Whoops. Meant `len_ + 1`. `capacity_ + 1` would guarantee we allocate on every call to `set()`